sys/proc.c

```diff
@@ -53,4 +63,142 @@ proc_t *proc_find(pid_t pid) {
+
+  /* Record some process statistics that will stay maintained in zombie
+     state. */
+  p->p_exitstatus = exitstatus;
+
+  /* Turn the process into a zombie. */
+  mtx_lock(&all_proc_list_mtx);
+  TAILQ_REMOVE(&all_proc_list, p, p_all);
+  mtx_unlock(&all_proc_list_mtx);
+
+  p->p_state = PRS_ZOMBIE;
+
+  mtx_lock(&zombie_proc_list_mtx);
+  TAILQ_INSERT_TAIL(&zombie_proc_list, p, p_zombie);
+  mtx_unlock(&zombie_proc_list_mtx);
+
+  /* Notify parent possibly waiting for this process to become zombie that it is
+     now. */
+  if (p->p_parent)
+    sleepq_broadcast(&p->p_parent->p_children);
+
+  mtx_unlock(&p->p_lock);
+
+  /* This thread is the last one in the process to exit. */
+  thread_exit();
+}
+
+/* These functions aren't very useful, but they clean up code layout by
+   splitting multiple levels of nested loops */
+static proc_t *proc_find_child(proc_t *p, pid_t pid) {
+  assert(mtx_owned(&p->p_lock));
+  proc_t *child;
+  TAILQ_FOREACH (child, &p->p_children, p_child) {
+    if (child->p_pid == pid)
+      return child;
+  }
+  return NULL;
+}
+
+/* Note: Returned child is locked. */
+static proc_t *proc_find_zombiechild(proc_t *p) {
+  assert(mtx_owned(&p->p_lock));
+  proc_t *child;
+  TAILQ_FOREACH (child, &p->p_children, p_child) {
+    mtx_lock(&child->p_lock);
+    if (child->p_state == PRS_ZOMBIE)
+      return child;
+    mtx_unlock(&child->p_lock);
+  }
+  return NULL;
+}
+
+int do_waitpid(pid_t pid, int *status, int options) {
+  /* We don't have a concept of process groups yet. */
+  if (pid < -1 || pid == 0)
+    return -ENOTSUP;
+
+  thread_t *td = thread_self();
+  proc_t *p = td->td_proc;
+  assert(p);
+
+  proc_t *child = NULL;
+
+  if (pid == -1) {
+    mtx_lock(&p->p_lock);
+    while (1) {
+
+      /* Search for any zombie children. */
+      child = proc_find_zombiechild(p);
+      if (child)
+        break;
+
+      /* No zombie child was found. */
+
+      mtx_unlock(&p->p_lock);
+      if (options & WNOHANG)
+        return -ECHILD;
+
+      /* Wait until a child changes state. */
+      sleepq_wait(&p->p_children, "child process");
+      mtx_lock(&p->p_lock);
+    }
+    mtx_unlock(&p->p_lock);
+  } else {
+    /* Wait for a particular child. */
+    mtx_lock(&p->p_lock);
+    child = proc_find_child(p, pid);
+    mtx_unlock(&p->p_lock);
+
+    if (!child) /* No such process, or the process is not a child. */
+      return -ECHILD;
+
+    mtx_lock(&child->p_lock);
+    while (child->p_state != PRS_ZOMBIE) {
+      mtx_unlock(&child->p_lock);
+      if (options & WNOHANG)
+        return 0;
+      sleepq_wait(&p->p_children, "child process");
+      mtx_lock(&child->p_lock);
+    }
+  }
+
+  /* Child is now a zombie. Gather its data, cleanup & free. */
+  assert(mtx_owned(&child->p_lock));
+  *status = child->p_exitstatus;
+  int retval = child->p_pid;
+
+  mtx_lock(&zombie_proc_list_mtx);
+  TAILQ_REMOVE(&zombie_proc_list, child, p_zombie);
+  mtx_unlock(&zombie_proc_list_mtx);
+
+  kfree(M_PROC, child);
```

Comment:
Hmm... initially I thought there could be a situation where more than one process waits for another process to finish. But apparently you can only wait for direct descendants (i.e. children) processes. Of course the question reemerges in presence of multiple threads, but that's not a thing for now.